### PR TITLE
[TTS][refactor] Part 4 - nn.collections.tts.g2p.modules

### PR DESCRIPTION
**Pull request by XuesongYang**
# What does this PR do ?

Task 4: break down `nemo/collections/tts/g2p/modules.py` and save to `nemo/collections/tts/g2p/models`. The “Modules” concept used by g2p is misleading in comparison to NeMo standards, i.e. modules usually mean `torch.nn.Module` or `NeuralModule`. We should remove any confusion. According to several offline discussion, we finalized below changes,
```
- nemo/collections/tts/g2p/data
  - ctc.py
  - heteronym_classification.py
  - t5.py
- nemo/collections/tts/g2p/models
  - base.py # only contains BaseG2p class in this file.
  - zh_cn_pinyin.py  # move nemo/collections/tts/g2p/modules.py::ChineseG2p class here.
  - en_us_arpabet.py  # move nemo/collections/tts/g2p/modules.py::EnglishG2p class here.
  - i18n_ipa.py  # move nemo/collections/tts/g2p/modules.py::IPAG2P class here. It is designed to support any language when specifying a locale code.
  - ctc.py
  - heteronym_classification.py
  - t5.py
- nemo/collections/tts/g2p/utils.py

- added deprecation notice for nemo/collections/tts/torch/g2ps.py
- renamed nemo/collections/tts/data/tts_dataset.py -> nemo/collections/tts/data/dataset.py
```

```
$ pytest --nightly tests/collections/tts/models/
A valid `test_data.tar.gz` test archive (10445891B) found in the `/home/xueyang/workspace/NeMo/tests/.data` folder.
Setting numba compat : True
===================================================== test session starts ======================================================
platform linux -- Python 3.8.13, pytest-7.1.3, pluggy-1.0.0 -- /home/xueyang/miniconda3/envs/nemo/bin/python
cachedir: .pytest_cache
rootdir: /home/xueyang/workspace/NeMo, configfile: pyproject.toml
plugins: hydra-core-1.2.0, anyio-3.6.1
collected 26 items                                                                                                             

tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner] PASSED                               [  3%]
tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner_ipa] PASSED                           [  7%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch] PASSED                                  [ 11%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_ipa] PASSED                              [ 15%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_multispeaker] PASSED                     [ 19%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_2102] PASSED [ 23%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_2210] PASSED [ 26%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_multispeaker_5] PASSED                   [ 30%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_es_fastpitch_multispeaker] PASSED                     [ 34%]
tests/collections/tts/models/test_fastpitch.py::test_inference[tts_zh_fastpitch_sfspeech] PASSED                         [ 38%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifigan] PASSED                                      [ 42%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixertts] PASSED                       [ 46%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixerttsx] PASSED                      [ 50%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifitts_hifigan_ft_fastpitch] PASSED                 [ 53%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2102] PASSED   [ 57%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2210] PASSED   [ 61%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hui_hifigan_ft_fastpitch_multispeaker_5] PASSED      [ 65%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_es_hifigan_ft_fastpitch_multispeaker] PASSED            [ 69%]
tests/collections/tts/models/test_hifigan.py::test_inference[tts_zh_hifigan_sfspeech] PASSED                             [ 73%]
tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixertts] PASSED                                 [ 76%]
tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixerttsx] PASSED                                [ 80%]
tests/collections/tts/models/test_tacotron2.py::test_inference[tts_en_tacotron2] PASSED                                  [ 84%]
tests/collections/tts/models/test_univNet.py::test_inference[tts_en_lj_univnet] PASSED                                   [ 88%]
tests/collections/tts/models/test_univNet.py::test_inference[tts_en_libritts_univnet] PASSED                             [ 92%]
tests/collections/tts/models/test_vits.py::test_inference[tts_en_lj_vits] PASSED                                         [ 96%]
tests/collections/tts/models/test_waveGlow.py::test_inference[tts_en_waveglow_88m] PASSED                                [100%]

====================================================== slowest durations =======================================================
26.97s setup    tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixerttsx]
25.43s setup    tests/collections/tts/models/test_tacotron2.py::test_inference[tts_en_tacotron2]
25.30s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_ipa]
24.99s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_multispeaker]
24.87s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch]
24.55s setup    tests/collections/tts/models/test_vits.py::test_inference[tts_en_lj_vits]
23.77s setup    tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner]
23.56s setup    tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixertts]
22.72s setup    tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner_ipa]
21.13s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_multispeaker_5]
20.04s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_22
10]
16.39s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_21
02]
2.42s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifigan]
2.39s setup    tests/collections/tts/models/test_waveGlow.py::test_inference[tts_en_waveglow_88m]
2.38s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_es_fastpitch_multispeaker]
2.32s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixertts]
2.31s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixerttsx]
1.81s call     tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixerttsx]
1.81s call     tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner]
1.56s setup    tests/collections/tts/models/test_fastpitch.py::test_inference[tts_zh_fastpitch_sfspeech]
1.40s setup    tests/collections/tts/models/test_univNet.py::test_inference[tts_en_libritts_univnet]
1.04s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hui_hifigan_ft_fastpitch_multispeaker_5]
0.99s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2210]
0.98s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifitts_hifigan_ft_fastpitch]
0.98s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_es_hifigan_ft_fastpitch_multispeaker]
0.97s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2102]
0.97s setup    tests/collections/tts/models/test_hifigan.py::test_inference[tts_zh_hifigan_sfspeech]
0.91s setup    tests/collections/tts/models/test_univNet.py::test_inference[tts_en_lj_univnet]
0.49s call     tests/collections/tts/models/test_vits.py::test_inference[tts_en_lj_vits]
0.47s call     tests/collections/tts/models/test_tacotron2.py::test_inference[tts_en_tacotron2]
0.46s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_zh_fastpitch_sfspeech]
0.28s call     tests/collections/tts/models/test_waveGlow.py::test_inference[tts_en_waveglow_88m]
0.07s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_multispeaker_5]
0.07s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_210
2]
0.07s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_de_fastpitch_singleSpeaker_thorstenNeutral_221
0]
0.06s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch]
0.06s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_ipa]
0.06s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_en_fastpitch_multispeaker]
0.05s call     tests/collections/tts/models/test_mixerTTS.py::test_inference[tts_en_lj_mixertts]
0.05s call     tests/collections/tts/models/test_aligner.py::test_inference[tts_en_radtts_aligner_ipa]
0.03s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifigan]
0.03s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2102]
0.03s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_hifitts_hifigan_ft_fastpitch]
0.03s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hui_hifigan_ft_fastpitch_multispeaker_5]
0.02s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_es_hifigan_ft_fastpitch_multispeaker]
0.02s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixertts]
0.02s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_en_lj_hifigan_ft_mixerttsx]
0.02s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_de_hifigan_singleSpeaker_thorstenNeutral_2210]
0.02s call     tests/collections/tts/models/test_hifigan.py::test_inference[tts_zh_hifigan_sfspeech]
0.02s call     tests/collections/tts/models/test_fastpitch.py::test_inference[tts_es_fastpitch_multispeaker]
0.02s call     tests/collections/tts/models/test_univNet.py::test_inference[tts_en_lj_univnet]
0.01s call     tests/collections/tts/models/test_univNet.py::test_inference[tts_en_libritts_univnet]

(26 durations < 0.005s hidden.  Use -vv to show these durations.)
================================================ 26 passed in 307.59s (0:05:07) ================================================
```

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/collections/asr/test_hybrid_asr_tts_models.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
 from nemo.collections.tts.models import FastPitchModel
 
 
+@pytest.mark.pleasefixme
```
Comment (XuesongYang) [inline tests/collections/asr/test_hybrid_asr_tts_models.py:28]: @artbataev for vis.

File: tests/collections/asr/test_hybrid_asr_tts_models.py [hunk 2]
```diff
@@ -116,6 +117,7 @@ def asr_tts_ctc_bpe_model(asr_model_ctc_bpe_config, fastpitch_model_path):
     return model
 
 
+@pytest.mark.pleasefixme
```
Comment (XuesongYang) [inline tests/collections/asr/test_hybrid_asr_tts_models.py:120]: @artbataev for vis.

File: tests/collections/tts/test_tts_exportables.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 from nemo.utils.app_state import AppState
 
 
+@pytest.mark.pleasefixme
```
Comment (XuesongYang) [inline tests/collections/tts/test_tts_exportables.py:25]: @borisfom for vis.

File: tests/collections/tts/test_tts_exportables.py [hunk 4]
```diff
@@ -60,6 +61,7 @@ def radtts_model():
 
 
 class TestExportable:
+    @pytest.mark.pleasefixme
```
Comment (XuesongYang) [inline tests/collections/tts/test_tts_exportables.py:64]: @borisfom for vis.

**Review by artbataev — CHANGES_REQUESTED** [review]
Please, do not make breaking changes.
This will significantly impact the teamwork and make our users and contributors unhappy.
This also violates NeMo's policy regarding breaking changes.

I think some good solutions exist for this problem:
1. If I understand correctly, you moved the classes from `nemo.collections.tts.g2p` to `nemo.collections.common.g2p`. So, it is possible to preserve compatibility by allowing them to be imported from the old modules
e.g. in `nemo.collections.tts.g2p.modules.IPAG2P' use 
```python
from nemo.collections.common.g2p.i18n_ipa import IpaG2p as IPAG2P

__all__ = [“IPAG2P”]
```
and so on.

Or use a **better solution with inheritance + deprecation warning**
```python
from nemo.collections.common.g2p.i18n_ipa import IpaG2p as NewIpaG2p
from nemo.utils.decorators.deprecated import deprecated

@deprecated(...)
class IPAG2P(NewIpaG2p):
    ...
```

2. If it is impossible to apply the first solution, provide a guide + scripts to convert existing checkpoints to make them compatible with the new version.

**Comment by titu1994** [comment]
Agreed, do not make backward compatibility breaking changes without following NeMo guidelines and appropriate release timeline for such changes (1-2 months of deprecated label, followed by removal). 

It doesn't seem you need to break anything jf you forward imports correctly, ASR does something similar - see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/features.py#L39

Plus avoid adding dependencies to commons that are optional dependencies - difficult to install things like apex and numba should avoid commons as much as possible so as not to impose restrictions on other domains.

**Comment by XuesongYang** [comment]
converted this PR to a draft for improvements.

**Comment by XuesongYang** [comment]
I added safeguard for TTS model inference: https://github.com/NVIDIA/NeMo/pull/6129.

**Review by redoctopus — COMMENTED** [review]
Are we sure we want to move G2P into `common`? I thought part of the reason it was moved back to the `tts` collection was because no other parts of NeMo actually use the G2P classes. Would it be clearer to rename the files in-place or move them to somewhere else in the TTS collection?

@ekmb @rlangman any thoughts?

**Comment by XuesongYang** [comment]
> Are we sure we want to move G2P into `common`? I thought part of the reason it was moved back to the `tts` collection was because no other parts of NeMo actually use the G2P classes. Would it be clearer to rename the files in-place or move them to somewhere else in the TTS collection?

G2P is a common tool for phoneme-based ASR, although recent research mostly focuses on sub-word units. So that's why I am thinking to promote g2p under `nemo.collections.common.g2p`. 

An alternative place to host g2p is under `nemo.collections.g2p` in parallel with `asr`, `nlp`, and `tts`. Then we can maintain similar sub-directory as tts/asr, and move g2p models under `nemo.collections.g2p.models`

**Comment by redoctopus** [comment]
> G2P is a common tool for phoneme-based ASR, although recent research mostly focuses on sub-word units. So that's why I am thinking to promote g2p under nemo.collections.common.g2p.

We don't have this yet, right? So this would be a forward-looking change? That's fine with me, as long as we're relatively certain we'll follow-through and not just have to move it back into TTS.

@titu1994 and @ekmb to solicit opinions about where to put these.

**Comment by ekmb** [comment]
> > G2P is a common tool for phoneme-based ASR, although recent research mostly focuses on sub-word units. So that's why I am thinking to promote g2p under nemo.collections.common.g2p.
> 
> We don't have this yet, right? So this would be a forward-looking change? That's fine with me, as long as we're relatively certain we'll follow-through and not just have to move it back into TTS.
> 
> @titu1994 and @ekmb to solicit opinions about where to put these.

I would keep g2p in the TTS along with neural G2P models unless other collections use it. Rule-based G2P development (if any) will be happening in the NeMo-text-processing repo.
Could you add a subfolder for g2p models to separate them from TTS models (https://github.com/NVIDIA/NeMo/tree/refactor-tts-g2p-module/nemo/collections/tts)/models/?

**Comment by XuesongYang** [comment]
> > > > G2P is a common tool for phoneme-based ASR, although recent research mostly focuses on sub-word units. So that's why I am thinking to promote g2p under nemo.collections.common.g2p.
> > > 
> > > 
> > > We don't have this yet, right? So this would be a forward-looking change? That's fine with me, as long as we're relatively certain we'll follow-through and not just have to move it back into TTS.
> > > @titu1994 and @ekmb to solicit opinions about where to put these.
> > 
> > 
> > I would keep g2p in the TTS along with neural G2P models unless other collections use it. Rule-based G2P development (if any) will be happening in the NeMo-text-processing repo. Could you add a subfolder for g2p models to separate them from TTS models (https://github.com/NVIDIA/NeMo/tree/refactor-tts-g2p-module/nemo/collections/tts)/models/?
> 
> The G2P in NeMo (non-neural modules) actually are rule-based ones. They should be merged into nemo-text-processing repo in the future I guess. Since most of you guys suggested moving g2p under tts folder, then two possible structures are shown below, which one do you guys prefer? @ekmb @redoctopus @rlangman. **Option 1**
> 
> ```
> nemo/collections/tts/models
> - base.py # move BaseG2p class into this file.
> - ...
> - g2p_en_us_arpabet.py
> - g2p_i18n_ipa.py
> - g2p_zh_cn_pinyin.py
> - g2p_ctc.py
> - g2p_heteronym_classification.py
> - g2p_t5.py
> - ...
> ```
> 
> **Option 2**
> 
> ```
> nemo/collections/tts/models/g2p
> - base.py # only contains BaseG2p class in this file.
> - en_us_arpabet.py
> - i18n_ipa.py
> - zh_cn_pinyin.py
> - ctc.py
> - heteronym_classification.py
> - t5.py
> ```

Discussed with folks offline, we finally decided to proceed with the below option,
```
- nemo/collections/tts/g2p/data
  - ctc.py
  - heteronym_classification.py
  - t5.py
- nemo/collections/tts/g2p/models
  - base.py # only contains BaseG2p class in this file.
  - en_us_arpabet.py
  - i18n_ipa.py
  - zh_cn_pinyin.py
  - ctc.py
  - heteronym_classification.py
  - t5.py
- nemo/collections/tts/g2p/utils.py
```

**Review by redoctopus — APPROVED** [review]
LGTM!

**Pull request merged**